src/libDirectoryService/Coinbase.cpp

```diff
@@ -390,75 +398,51 @@ void DirectoryService::InitCoinbase() {
   DiagnosticDataCoinbase entry = {
-      node_count,  sig_count,        lookup_count, total_reward + base_reward,
+      node_count,  sig_count,        lookup_count, total_reward,
       base_reward, base_reward_each, lookupReward, reward_each_lookup,
       nodeReward,  reward_each,      balance_left, PubKey(PrivKey()),
       Address()};
-  for (const auto& shard : m_coinbaseRewardees[m_mediator.m_currentEpochNum]) {
-    if (count == shardIndex) {
-      uint16_t rdm_index = lastBlockHash % shard.second.size();
-      uint16_t rdm_index_copy = rdm_index;
-      if (GUARD_MODE) {
-        while (Guard::GetInstance().IsNodeInDSGuardList(
-                   shard.second.at(rdm_index)) ||
-               Guard::GetInstance().IsNodeInShardGuardList(
-                   shard.second.at(rdm_index))) {
-          rdm_index = (rdm_index + 1) % shard.second.size();
-          if (rdm_index == rdm_index_copy) {
-            LOG_GENERAL(WARNING, "Going into infinite loop");
-            breakLoop = true;
-            break;
-          }
-        }
-      }
-      if (breakLoop) {
-        break;
-      }
+  if (nonGuard.empty()) {
+    LOG_GENERAL(WARNING, "No non-guard found, skip LuckyDraw");
+    return;
+  }
 
-      const Address& winnerAddr =
-          Account::GetAddressFromPublicKey(shard.second.at(rdm_index));
-      LOG_GENERAL(INFO, "Lucky draw winner: " << winnerAddr);
-      if (!AccountStore::GetInstance().UpdateCoinbaseTemp(
-              winnerAddr, coinbaseAddress, balance_left)) {
-        LOG_GENERAL(WARNING, "Could not reward lucky draw!");
-      }
+  uint16_t luckyIndex = lastBlockHash % nonGuard.size();
 
-      // Only log reward for my self so can find out the reward of mine in state
-      // log
-      if (winnerAddr == myAddr) {
-        LOG_STATE("[REWARD][" << setw(15) << left
-                              << m_mediator.m_selfPeer.GetPrintableIPAddress()
-                              << "][" << m_mediator.m_currentEpochNum << "]["
-                              << balance_left << "] lucky draw");
-      }
+  auto const& luckyAddr = nonGuard.at(luckyIndex);
 
-      entry.luckyDrawWinnerKey = shard.second.at(rdm_index);
-      entry.luckyDrawWinnerAddr = winnerAddr;
-      StoreCoinbaseInDiagnosticDB(entry);
+  LOG_GENERAL(INFO, "Lucky draw winner: " << luckyAddr);
+  if (!AccountStore::GetInstance().UpdateCoinbaseTemp(
+          luckyAddr, coinbaseAddress, balance_left)) {
+    LOG_GENERAL(WARNING, "Could not reward lucky draw!");
+  }
 
-      return;
-    } else {
-      ++count;
-    }
+  // Only log reward for my self so can find out the reward of mine in state
+  // log
+  if (luckyAddr == myAddr) {
+    LOG_STATE("[REWARD][" << setw(15) << left
+                          << m_mediator.m_selfPeer.GetPrintableIPAddress()
+                          << "][" << m_mediator.m_currentEpochNum << "]["
+                          << balance_left << "] lucky draw");
   }
 
   StoreCoinbaseInDiagnosticDB(entry);
```

Comment:
Set `entry.luckyDrawWinnerAddr = luckyAddr` here
Note that `entry.luckyDrawWinnerKey` cannot be known anymore, but should be ok